### PR TITLE
Split A/B Test: Return all possible cache keys

### PR DESCRIPTION
**Pull request by kapoli**
* Modify `get_all_possible_variations` to return all possible permutations of a/b test variations

### CODE REVIEW

File: Gemfile [hunk 1]
```diff
@@ -2,3 +2,4 @@ source 'https://rubygems.org'
 
 # Specify your gem's dependencies in split_cacheable.gemspec
 gemspec
+gem 'split', '1.0.0'
```
Comment (pjambet) [inline Gemfile]: This should be in the [gemspec](https://github.com/harrystech/split_cacheable/blob/master/split_cacheable.gemspec#L21) file

**Comment by kapoli** [comment]
@pjambet - updated gemspec instead of gemfile

File: split_cacheable.gemspec [hunk 2]
```diff
@@ -18,7 +18,7 @@ Gem::Specification.new do |spec|
   spec.test_files    = spec.files.grep(%r{^(test|spec|features)/})
   spec.require_paths = ["lib"]
 
-  spec.add_dependency "split", ">=  0.7.1"
+  spec.add_dependency "split", "1.0.0"
```
Comment (danielschwartz) [inline split_cacheable.gemspec]: You will want this to be an "around" dep, not a strict dep. My guess is we support 1.0.x as well.

File: Gemfile [hunk 3]
```diff
@@ -1,4 +1,4 @@
 source 'https://rubygems.org'
 
 # Specify your gem's dependencies in split_cacheable.gemspec
-gemspec
+gemspec
```
Comment (pjambet) [inline Gemfile]: Can you remove this change from the commit?

File: lib/split_cacheable.rb [hunk 4]
```diff
@@ -83,8 +83,13 @@ def get_all_possible_variations
                     test_variations[0].unshift(DEFAULT_KEY)
                     return test_variations[0]
                 else
-                    first_test = test_variations.shift
-                    all_variations = first_test.product(*test_variations).map{|a| a.join("/")}
+                    all_variations = Array.new
```
Comment (pjambet) [inline lib/split_cacheable.rb]: We usually initialize arrays with `all_variations = []`

File: lib/split_cacheable.rb [hunk 5]
```diff
@@ -83,8 +83,13 @@ def get_all_possible_variations
...
+                    (1..test_variations.length).each do |i|
```
Comment (pjambet) [inline lib/split_cacheable.rb]: I think it'd be a little bit better to use `test_variations .each_with_index` or `.each.with_index`

File: lib/split_cacheable.rb [hunk 6]
```diff
@@ -83,8 +83,13 @@ def get_all_possible_variations
...
+                        sets = test_variations.combination(i).to_a
```
Comment (pjambet) [inline lib/split_cacheable.rb]: Why do you need to use `to_a`?

**Comment by pjambet** [comment]
I think the commit message, and the name of the branch are wrong, the cache keys are not technically tied to redis, it could be any cache store set in Rails.cache.

**Comment by kapoli** [comment]
@pjambet - Split itself using redis as a datastore

File: spec/cacheable_spec.rb [hunk 7]
```diff
@@ -1,7 +1,7 @@
 require 'spec_helper'
 
 describe Split::Cacheable do
-    ALL_VARIATIONS = [Split::Cacheable::Adapter::DEFAULT_KEY, "test_1/old/test_2/left", "test_1/old/test_2/right", "test_1/new/test_2/left", "test_1/new/test_2/right"]
+    ALL_VARIATIONS = [Split::Cacheable::Adapter::DEFAULT_KEY, "test_1/old", "test_1/new", "test_2/left", "test_2/right", "test_1/old/test_2/left", "test_1/old/test_2/right", "test_1/new/test_2/left", "test_1/new/test_2/right"]
```
Comment (pjambet) [inline spec/cacheable_spec.rb]: Shouldn't we add a test case where there are two tests on the same action, but with opposite `if`?

Comment (kapoli) [inline spec/cacheable_spec.rb]: @pjambet just added a spec on a test controller with 2 tests on the same action

File: lib/split_cacheable.rb [hunk 8]
```diff
@@ -83,8 +83,13 @@ def get_all_possible_variations
                     test_variations[0].unshift(DEFAULT_KEY)
                     return test_variations[0]
                 else
-                    first_test = test_variations.shift
-                    all_variations = first_test.product(*test_variations).map{|a| a.join("/")}
+                    all_variations = []
+                    (1..test_variations.length).each do |i|
```
Comment (pjambet) [inline lib/split_cacheable.rb]: As I said in a previous comment that is now hidden for some unknown reason (I guess this line has technically chenged, thanks gh!):
I think it'd be a little bit better to use test_variations.each_with_index or .each.with_index

File: lib/split_cacheable.rb [hunk 9]
```diff
@@ -83,8 +83,13 @@ def get_all_possible_variations
...
+                        sets = test_variations.combination(i)
+                        sets.each do |set|
```
Comment (pjambet) [inline lib/split_cacheable.rb]: you don't need the `sets` variable

File: spec/spec_helper.rb [hunk 10]
```diff
@@ -80,4 +80,19 @@ def index
...
+end
+
+class TestControllerWithRequestAndMultiTestOnSameAction < TestControllerWithRequest
+    cacheable_ab_test :test_1, :only => :index
+    cacheable_ab_test :test_2, :only => :index
+    def index
+        
+    end
+
+    def request(ua = 'Mozilla/5.0 (Macintosh; U; Intel Mac OS X 10_6_6; de-de) AppleWebKit/533.20.25 (KHTML, like Gecko) Version/5.0.4 Safari/533.20.27')
```
Comment (pjambet) [inline spec/spec_helper.rb]: I don' t think you need this method, it is already inherited from `Split::ExperimentCatalog`

**Comment by kapoli** [comment]
Hey @pjambet, I updated the spec

**Comment by pjambet** [comment]
Cool. Looks good to me.
As I said offline, I still think it's a little bit weird that we generate all the combination when we _should_ never have an actual cache key composed of multiple variations.
But I guess it's better to generate too many keys rather than not enough, as we used to.

@danielschwartz As you wrote this gem, can you give it a final look.

**Comment by danielschwartz** [comment]
Yeah, I spoke about this with @kapoli when she was starting. I agree that it's subpar to generate all possible hash keys when we may not have activated a large percentage of them, but given that the :if blocks can execute any arbitrary code, I don't see a better way of doing this.

**Pull request merged**